parser/src/parser/file.rs

```diff
@@ -520,8 +520,16 @@ impl ParserContext<'_> {
             None
         };
 
-        // Parse the function body.
-        let block = self.parse_block()?;
+        // Parse the function body if it exists and we are not an internal parse.
+        // If it is internal the body definition allowed to be None.
+        let (block, span) = if self.peek()?.token == Token::Semicolon && self.internal {
```

Comment:
This `internal` business, and similar semantic restrictions, is better handled outside the parser as an "AST validation" pass or during lowering to the ASG. Or at least, you could have parser recovery for `if !self.internal { ... }`.

Reply:
I'm fine with handling it another way. But not sure how we would get to AST validation if we can't parse it first?

Could you explain the use of recovery here? It's not a feature we want to expose to users anyways so I don't see a case for it?

Reply:
> But not sure how we would get to AST validation if we can't parse it first?

You can parse `fn foo(...);` even if it is not internal.

> Could you explain the use of recovery here? It's not a feature we want to expose to users anyways so I don't see a case for it?

It's a reasonable for the user to try something like `fn foo();` and then having recovery is helpful. See for example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1ddc0668d5a42f6bdf79d706bef7ad44
This is handled in rustc's AST validation:

https://github.com/rust-lang/rust/blob/68568dcb8ff49a3d70f4cc2d9215b5753d088738/compiler/rustc_ast_passes/src/ast_validation.rs#L1155-L1161

Reply:
Ah, I see you're suggesting allowing at the parsing level always, then validating at the AST level. I'm okay with this. What do you think, @acoglio?

Reply:
Yes, I thought that in fact we would be extending the ABNF grammar to allow semicolons, in alternative to blocks, in function declarations, so that the stdlib files that use this feature are grammatically valid. Then we can have a static semantic check that disallows user-defined functions. The change to the ABNF grammar was described in the RFC: https://github.com/AleoHQ/leo/blob/master/docs/rfc/010-native-functions.md#proposal